### PR TITLE
macOS osx/ios dynamic framework support

### PR DESCRIPTION
**Pull request by halx99**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -65,6 +65,8 @@ option(ALSOFT_INSTALL_EXAMPLES "Install example programs (alplay, alstream, ...)
 option(ALSOFT_INSTALL_UTILS "Install utility programs (openal-info, alsoft-config, ...)" ON)
 option(ALSOFT_UPDATE_BUILD_VERSION "Update git build version info" ON)
 
+option(ALSOFT_OSX_FRAMEWORK "Build as macOS framework support" ON)
```
Comment (alexey-lysiuk) [inline CMakeLists.txt:68]: Could you please disable this by default?

File: CMakeLists.txt [hunk 2]
```diff
@@ -174,10 +176,12 @@ endif()
...
-check_library_exists(log __android_log_print "" HAVE_LIBLOG)
-if(HAVE_LIBLOG)
-    set(EXTRA_LIBS log ${EXTRA_LIBS})
-    set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} log)
+if(ANDROID)
+    check_library_exists(log __android_log_print "" HAVE_LIBLOG)
+    if(HAVE_LIBLOG)
+        set(EXTRA_LIBS log ${EXTRA_LIBS})
+        set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} log)
+    endif()
```
Comment (alexey-lysiuk) [inline CMakeLists.txt:184]: Just out of curiosity, how is this change related to a framework creation?

File: CMakeLists.txt [hunk 3]
```diff
@@ -1221,6 +1225,17 @@ else()
...
     target_link_libraries(${IMPL_TARGET} PRIVATE common ${LINKER_FLAGS} ${EXTRA_LIBS} ${MATH_LIB})
+
+    if(APPLE AND ALSOFT_OSX_FRAMEWORK)
+        set_target_properties(${IMPL_TARGET} PROPERTIES
+            FRAMEWORK TRUE
+            FRAMEWORK_VERSION C
+            MACOSX_FRAMEWORK_IDENTIFIER org.opena-soft.openal
+            MACOSX_FRAMEWORK_INFO_PLIST ${OpenAL_SOURCE_DIR}/info.plist
+            PUBLIC_HEADER include/AL/alc.h include/AL/al.h include/AL/alext.h 
+            XCODE_ATTRIBUTE_CODE_SIGN_IDENTITY "iPhone Developer"
```
Comment (alexey-lysiuk) [inline CMakeLists.txt:1236]: Is this needed for command line based project generated by CMake? Xcode project should work fine without it.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1221,6 +1225,17 @@ else()
         set_target_properties(${IMPL_TARGET} PROPERTIES PREFIX "")
     endif()
     target_link_libraries(${IMPL_TARGET} PRIVATE common ${LINKER_FLAGS} ${EXTRA_LIBS} ${MATH_LIB})
+
+    if(APPLE AND ALSOFT_OSX_FRAMEWORK)
+        set_target_properties(${IMPL_TARGET} PROPERTIES
+            FRAMEWORK TRUE
+            FRAMEWORK_VERSION C
+            MACOSX_FRAMEWORK_IDENTIFIER org.opena-soft.openal
```
Comment (alexey-lysiuk) [inline CMakeLists.txt:1233]: L character is missing in the domain name.

File: CMakeLists.txt (hunk 1 above)
Comment (halx99) [reply CMakeLists.txt:68]: Yeah, should disable by default

File: CMakeLists.txt (hunk 2 above)
Comment (halx99) [reply CMakeLists.txt:184]: please refer to: https://github.com/kcat/openal-soft/pull/452

File: CMakeLists.txt (hunk 4 above)
Comment (halx99) [reply CMakeLists.txt]: fixed

File: CMakeLists.txt (hunk 2 above)
Comment (alexey-lysiuk) [reply CMakeLists.txt:184]: Still not sure how this Android-specific thing affects creation of a framework.

Comment (halx99) [reply CMakeLists.txt:184]: it will lead link issue, -llog not exist

Comment (alexey-lysiuk) [reply CMakeLists.txt:184]: `HAVE_LIBLOG` should not be defined after `check_library_exists()` call if there is no such library.

Comment (halx99) [reply CMakeLists.txt:184]: but it doesn't work, you can try my fork on macOS 11.15, xcode11, cmake-3.17+

File: CMakeLists.txt (hunk 3 above)
Comment (halx99) [reply CMakeLists.txt:1236]:
```sh
cd build
cmake -GXcode .. -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_REQUIRE_COREAUDIO=ON  -DALSOFT_EMBED_HRTF_DATA=YES -DLIBTYPE=SHARED -DALSOFT_UTILS=OFF -DALSOFT_EXAMPLES=OFF -DALSOFT_INSTALL=OFF -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_OSX_FRAMEWORK=ON
# open OpenAL xcode project, then build the OpenAL framework
```

File: CMakeLists.txt (hunk 2 above)
Comment (alexey-lysiuk) [reply CMakeLists.txt:184]: Please define "doesn't work". If it was detection errors, we would like to know why did it happen.

Regardless of the actual issue, I cannot comprehend why this affects framework creation.
I built static and dynamic libraries with various versions of Xcode, and never had any issues with this.

File: CMakeLists.txt [hunk 5]
```diff
@@ -1221,6 +1225,17 @@ else()
...
+            MACOSX_FRAMEWORK_INFO_PLIST ${OpenAL_SOURCE_DIR}/info.plist
+            PUBLIC_HEADER include/AL/alc.h include/AL/al.h include/AL/alext.h
```
Comment (kcat) [inline CMakeLists.txt:1235]: Should this also have `include/AL/efx.h`, since `alext.h` includes it (and some code may try including it directly)?

**Comment by halx99** [comment]
PUBLIC_HEADER need to fix

**Comment by halx99** [comment]
@kcat This pr should be works fine and finished, the attachment is build by me, if you have any changes request, please tell me

[openal.framework.zip](https://github.com/kcat/openal-soft/files/5100861/openal.framework.zip)

build step:
```
# build device
cmake -S . -Bbuild -GXcode -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_REQUIRE_COREAUDIO=ON  -DALSOFT_EMBED_HRTF_DATA=YES -DLIBTYPE=SHARED -DALSOFT_UTILS=OFF -DALSOFT_EXAMPLES=OFF -DALSOFT_INSTALL=OFF -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_OSX_FRAMEWORK=ON -DCMAKE_TOOLCHAIN_FILE=make/ios.toolchain.cmake
cmake --build build --config Release --target OpenAL

# build simulator64
cmake -S . -Bbuild_x64 -GXcode -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_REQUIRE_COREAUDIO=ON  -DALSOFT_EMBED_HRTF_DATA=YES -DLIBTYPE=SHARED -DALSOFT_UTILS=OFF -DALSOFT_EXAMPLES=OFF -DALSOFT_INSTALL=OFF -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_OSX_FRAMEWORK=ON -DCMAKE_TOOLCHAIN_FILE=make/ios.toolchain.cmake -DPLATFORM=SIMULATOR64
cmake --build build_x64 --config Release --target OpenAL

# combine device & simulator dyn libs to on flat-file
lipo -create build/Release-iphoneos/openal.framework/openal build_x64/build/Release-iphoneos/openal.framework/openal /openal.framework/openal -output ./openal
```

File: info.plist [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    <key>CFBundleIdentifier</key>
+    <string>org.openal-soft.openal</string>
+    <key>CFBundleInfoDictionaryVersion</key>
+    <string>6.0</string>
+    <key>CFBundleName</key>
+    <string>${PROJECT_NAME}</string>
+    <key>CFBundlePackageType</key>
+    <string>APPL</string>
+    <key>CFBundleShortVersionString</key>
+    <string>1.20.1</string>
```
Comment (kcat) [inline info.plist]: Would it be possible to automatically set this from the `LIB_VERSION` or `LIB_*_VERSION` vars in cmake? Does this file recognize cmake variables?

Comment (halx99) [reply info.plist]: Let me check it

File: CMakeLists.txt [hunk 7]
```diff
@@ -1220,7 +1220,9 @@ else()
         set(RC_CONFIG resources/soft_oal.rc)
     endif()
 
-    add_library(${IMPL_TARGET} SHARED ${OPENAL_OBJS} ${ALC_OBJS} ${RC_CONFIG})
+    # !important: for osx framework public header works, the headers must be in project
+    set(TARGET_PUBLIC_HEADERS include/AL/al.h;include/AL/alc.h;include/AL/alext.h;include/AL/efx.h)
```
Comment (kcat) [inline CMakeLists.txt:1224]: These semicolons don't look proper. If they're supposed to be separate elements in a list, spaces would be better to use.

Comment (halx99) [reply CMakeLists.txt:1224]: Yeah, please see the latest commit, fixed

File: info.plist (hunk 6 above)
Comment (halx99) [reply info.plist]: I push a new commit to auto set the version string
Note: I already remove info.plst file since I use CMake provided template: ```MacOSXFrameworkInfo.plist.in```
see: https://cmake.org/cmake/help/v3.13/prop_tgt/MACOSX_FRAMEWORK_INFO_PLIST.html#prop_tgt:MACOSX_FRAMEWORK_INFO_PLIST

File: CMakeLists.txt (hunk 7 above)
Comment (halx99) [reply CMakeLists.txt:1224]: I have checked again, space is works since write as:
```cmake
PUBLIC_HEADER "${TARGET_PUBLIC_HEADERS}"
```
I will change later.

Comment (halx99) [reply CMakeLists.txt]: modified

File: CMakeLists.txt [hunk 8]
```diff
@@ -1229,12 +1229,21 @@ else()
...
+        else()
+            set(BUNDLE_VERSION 1)
+        endif()
         set_target_properties(${IMPL_TARGET} PROPERTIES
             FRAMEWORK TRUE
             FRAMEWORK_VERSION C
             MACOSX_FRAMEWORK_IDENTIFIER "org.openal-soft.openal"
             MACOSX_FRAMEWORK_SHORT_VERSION_STRING ${LIB_VERSION}
-            MACOSX_FRAMEWORK_BUNDLE_VERSION 1
+            MACOSX_FRAMEWORK_BUNDLE_VERSION ${BUNDLE_VERSION}
```
Comment (kcat) [inline CMakeLists.txt:1246]: From what [I'm](https://cmake.org/cmake/help/v3.15/prop_tgt/MACOSX_FRAMEWORK_INFO_PLIST.html) [reading](https://developer.apple.com/documentation/bundleresources/information_property_list/cfbundleversion), this should be a `major.minor.patch` version, such as from `LIB_VERSION`, rather than a commit count.

Comment (halx99) [reply CMakeLists.txt:1246]: Nope, from my mobile development experience, the CFBundleVersion should be a number, the means same with ```buildNumber```

Comment (halx99) [reply CMakeLists.txt:1246]: see the snapshot:
image

File: CMakeLists.txt (hunk 2 above)
Comment (kcat) [reply CMakeLists.txt:184]: Please look into this, too. If `-llog` doesn't exist, it should not try to use it and so won't be a link problem. It could only succeed if the function `__android_log_print` specifically exists along with `liblog`. Specifically checking for Android first would be redundant. Is the check succeeding for some reason? What does the cmake log (in `CMakeFiles/CMakeOutput.log`) say for the function check?

Comment (halx99) [reply CMakeLists.txt:184]: See the snapshot:
image

**Comment by halx99** [comment]
Note: the framework name may needs to rename, because apple system framework name is ```OpenAL```

**Comment by halx99** [comment]
see the CMake command line output, even through I specific library find path wen use ```find_library```, still find apple system openal.framework:

CMake config code is:
```cmake
find_library(OPENAL_LIBRARY openal ${ALSOFT_FRAMEWORK_PATH})
# the var ALSOFT_FRAMEWORK_PATH is where the ```openal-soft's openal.framework``` at
```

```txt
-- before find OPENAL_LIBRARY=
-- OPENAL_LIBRARY=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS13.6.sdk/System/Library/Frameworks/openal.framework
```

**Comment by halx99** [comment]
Update: I link ```openal-soft's openal.framework``` with full path directly, and works fine:
image

**Comment by halx99** [comment]
so, does necessary rename framework name, @kcat what do you think about?

File: CMakeLists.txt (hunk 2 above)
Comment (kcat) [reply CMakeLists.txt:184]: `check_library_exists` caches its results, so if `HAVE_LIBLOG` was already set to true for some reason, it'll forever stay true and not be re-checked until the cache is cleared. What does `CMakeOutput.log` show for the `__android_log_print` check?

**Comment by kcat** [comment]
> so, does necessary rename framework name, @kcat what do you think about?

Since I don't think frameworks can be overridden in the same way plain shared libraries can, either at build-time or run-time, a different name may be appropriate. Though from what I see in `FindOpenAL.cmake`, it first lists `~/Library/Frameworks` as a search path before `/Library/Frameworks`. Does copying it to `~/Library/Frameworks/openal.framework` make it get selected over `/Library/Frameworks/openal.framework` (make sure to clear the project's cmake cache so it searches again)?

**Comment by halx99** [comment]
I think overrite system framework not good, because a non-system dynamic framework must as a embedded content of iphone app, and system framework not required

File: CMakeLists.txt (hunk 2 above)
Comment (halx99) [reply CMakeLists.txt:184]: Yeah, I clear the cache, I will upload two CMakeOutput.log later
1. CMake without toolchain ```ios.toolchain.cmake``` use iphone simulator sdk, the check should be fine, log not exist on ios
2. CMake with toolchain ```ios.toolchain.cmake`` use iphone device sdk, the check_library_exists not fine, will ```HAVE_LIBLOG``` will be ```true```

Comment (halx99) [reply CMakeLists.txt:184]: See the attachment file, it contains 2 file of CMake output for simulator_x64 and device_arm:

[CMakeOutputs.zip](https://github.com/kcat/openal-soft/files/5106631/CMakeOutputs.zip)

**Comment by halx99** [comment]
[CMakeOutputs.zip](https://github.com/kcat/openal-soft/files/5106631/CMakeOutputs.zip)

File: CMakeLists.txt (hunk 2 above)
Comment (halx99) [reply CMakeLists.txt:184]: Generate Command
* device arm: cmake -S . -GXcode -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_REQUIRE_COREAUDIO=ON  -DALSOFT_EMBED_HRTF_DATA=YES -DLIBTYPE=SHARED -DALSOFT_UTILS=OFF -DALSOFT_EXAMPLES=OFF -DALSOFT_INSTALL=OFF -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_OSX_FRAMEWORK=ON -DCMAKE_TOOLCHAIN_FILE=cmake/ios.toolchain.cmake -Bbuild_arm
* iphone simulator: cmake -S . -GXcode -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_REQUIRE_COREAUDIO=ON  -DALSOFT_EMBED_HRTF_DATA=YES -DLIBTYPE=SHARED -DALSOFT_UTILS=OFF -DALSOFT_EXAMPLES=OFF -DALSOFT_INSTALL=OFF -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_OSX_FRAMEWORK=ON -DCMAKE_TOOLCHAIN_FILE=cmake/ios.toolchain.cmake -Bbuild_arm

Comment (kcat) [reply CMakeLists.txt:184]: > Yeah, I clear the cache, I will upload two CMakeOutput.log later
> 
> 1. CMake without toolchain `ios.toolchain.cmake` use iphone simulator sdk, the check should be fine, log not exist on ios
> 2. CMake with toolchain `ios.toolchain.cmake` use iphone device sdk, the check_library_exists not fine, will  `HAVE_LIBLOG` will be `true`

I notice all `check_library_exists` checks seem to succeed; the others, libm, libdl, and librt, make sense to on a POSIX-based system, but liblog shouldn't. It looks like this may be at fault:
> ENABLE_STRICT_TRY_COMPILE: (1|0) Enables or disables strict try_compile() on all Check* directives (will run linker
>   to actually check if linking is possible). **Default 0** (false, will set CMAKE_TRY_COMPILE_TARGET_TYPE to STATIC_LIBRARY)

I don't know why it defaults to false, since a static library avoids any link-time error for missing symbols and dependent libraries, making checks for symbols and libraries like this succeed despite one (or both) not existing. It also seems to be a pretty useless option, since it just sets a variable that the user can set; rather than setting `ENABLE_STRICT_TRY_COMPILE=0` or `1`, set `CMAKE_TRY_COMPILE_TARGET_TYPE=STATIC_LIBRARY` or `EXECUTABLE`. The only difference is, the toolchain sets a different (and more error-prone) default on you.

So now there's two options. Tell all users they need to pass `-DENABLE_STRICT_TRY_COMPILE=1` to cmake when building for iPhone, or modify our local `ios.toolchain.cmake` to use a different default (or really, remove the option altogether because it can't not be set), which means we need to make sure to change it again whenever we update our local copy.

Comment (halx99) [reply CMakeLists.txt:184]: Hi, kcat
  I want to tell you, when I pass ```-DENABLE_STRICT_TRY_COMPILE=1```, everything check failed or not found

Comment (halx99) [reply CMakeLists.txt:184]: News: I install cmake-3.18.2, the check issue of cmake seems solved, I can confirm, previous cmake-3.18.1 doesn't works, then when use cmake-3.18.2, doesn't needs ```ios.toolchain.cmake```, the follow command can be execute succeed:
```
cmake -S . -GXcode -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_REQUIRE_COREAUDIO=ON -DALSOFT_EMBED_HRTF_DATA=YES -DLIBTYPE=SHARED -DALSOFT_UTILS=OFF -DALSOFT_EXAMPLES=OFF -DALSOFT_INSTALL=OFF -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_OSX_FRAMEWORK=ON -DCMAKE_SYSTEM_NAME=iOS -Bbuild_arm
```
The cmake-3.18.2 released 16 hours ago

Comment (kcat) [reply CMakeLists.txt:184]: > I want to tell you, when I pass `-DENABLE_STRICT_TRY_COMPILE=1`, everything check failed or not found

Can you provide `CMakefiles/CMakeError.log` from those failures?

> News: I install cmake-3.18.2, the check issue of cmake seems solved
> ...
> The cmake-3.18.2 released 16 hours ago

Way too new to depend on. iPhone shouldn't require a bleeding edge CMake to build for.

Comment (halx99) [reply CMakeLists.txt:184]: OK, let me install old version and try again

Comment (halx99) [reply CMakeLists.txt:184]: please see the attachment file:
[CMakeErrorAndOutput.zip](https://github.com/kcat/openal-soft/files/5107250/CMakeErrorAndOutput.zip)

**Comment by halx99** [comment]
[CMakeErrorAndOutput.zip](https://github.com/kcat/openal-soft/files/5107250/CMakeErrorAndOutput.zip)

**Comment by kcat** [comment]
```
Determining if the posix_memalign exist failed with the following output:
Change Dir: /Users/deal/dev/openal-soft/build_arm2/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/xcodebuild -project CMAKE_TRY_COMPILE.xcodeproj build -target cmTC_bf51d -configuration Debug -hideShellScriptEnvironment && Command line invocation:
    /Applications/Xcode.app/Contents/Developer/usr/bin/xcodebuild -project CMAKE_TRY_COMPILE.xcodeproj build -target cmTC_bf51d -configuration Debug -hideShellScriptEnvironment

User defaults from command line:
    HideShellScriptEnvironment = YES

Prepare build
note: Using legacy build system
=== BUILD TARGET cmTC_bf51d OF PROJECT CMAKE_TRY_COMPILE WITH CONFIGURATION Debug ===

Check dependencies
Code Signing Error: Bundle identifier is missing. cmTC_bf51d doesn't have a bundle identifier. Add a value for PRODUCT_BUNDLE_IDENTIFIER in the build settings editor.

** BUILD FAILED **
```
Would there be a way to disable code signing requirements for the `try_compile` tests, or set a dummy identifier to pacify the link?

**Comment by halx99** [comment]
xcode12+ 1k build script:
[build_ios.sh.zip](https://github.com/kcat/openal-soft/files/5581308/build_ios.sh.zip)
